drv/lpc55-swd/src/main.rs

```diff
@@ -1058,21 +1297,626 @@ impl ServerImpl {
+        }
+        Err(())
+    }
+
+    fn do_write_core_register(
+        &mut self,
+        register: u16,
+        value: u32,
+    ) -> Result<(), SpCtrlError> {
+        // C1.6 Debug system registers
+        let r = match register {
+            // R0-R12
+            0b0000000..=0b0001100
+
+            // LR - PSP
+            | 0b0001101..=0b0010010
+
+            // CONTROL/FAULTMASK/BASEPRI/PRIMASK
+            | 0b0010100
+
+            // FPCSR
+            | 0b0100001
+
+            // S0-S31
+            | 0b1000000..=0b1011111 => Ok::<u16, SpCtrlError>(register),
+            _ => {
+                ringbuf_entry!(Trace::InvalidRegisterWrite(register, value));
+                Err(SpCtrlError::InvalidCoreRegister)
+            }
+        }?;
+
+        self.write_single_target_addr(DCRDR, value)
+            .map_err(|_| SpCtrlError::Fault)?;
+        self.write_single_target_addr(DCRSR, r as u32 | (1u32 << 16))
+            .map_err(|_| SpCtrlError::Fault)?;
+
+        const RETRY_LIMIT: u32 = 10;
+        let mut limit = RETRY_LIMIT;
+        loop {
+            match self.dp_read_bitflags::<Dhcsr>() {
+                Ok(dhcsr) => {
+                    if dhcsr.is_regrdy() {
+                        // Trace retries used
+                        if limit != RETRY_LIMIT {
+                            ringbuf_entry!(Trace::LimitRemaining(limit));
+                        }
+                        return Ok(());
+                    }
+                    if limit == 0 {
+                        ringbuf_entry!(Trace::LimitRemaining(limit));
+                        return Err(SpCtrlError::Fault);
+                    }
+                    limit -= 1;
+                    hl::sleep_for(1);
+                }
+                Err(_) => return Err(SpCtrlError::Fault),
+            }
+        }
+    }
+
+    /// Measure the current SP Hubris Image.
+    /// The SP reset vector has been trapped and
+    /// the SP is halted.
+    fn do_measure_sp(&mut self) -> Result<[u8; 256 / 8], ()> {
+        // For Hubris on the STM32H7, The FWID includes 0xff padding to
+        // the end of the flash bank.
+
+        let mut magic = 0;
+        let mut total_image_size = 0;
+
+        // Sanity check the SP Hubris image
+        if self
+            .read_buf_from_addr(
+                endoscope::FLASH_BASE + sp::IMAGE_HEADER_MAGIC_OFFSET,
+                magic.as_bytes_mut(),
+            )
+            .is_err()
+            || magic != userlib::HEADER_MAGIC
+            || self
+                .read_buf_from_addr(
+                    endoscope::FLASH_BASE + sp::IMAGE_HEADER_LENGTH_OFFSET,
+                    total_image_size.as_bytes_mut(),
+                )
+                .is_err()
+            || !(sp::MIN_IMAGE_SIZE..(endoscope::FLASH_SIZE as usize))
+                .contains(&(total_image_size as usize))
+        {
+            ringbuf_entry!(Trace::SpHubrisFailsSanityCheck);
+            return Err(());
+        }
```

Comment:
What's the benefit to checking if the hubris image is valid? What is this going to look like from the higher levels?

Reply:
We don't need to have any Hubris sanity check here. It was reassuring at the start of this effort but it adds no value.

Reply:
deleted